symengine/fields.h

```diff
@@ -0,0 +1,320 @@
+
+    GaloisFieldDict &operator=(GaloisFieldDict &&other) SYMENGINE_NOEXCEPT
+    {
+        if (this != &other) {
+            dict_ = std::move(other.dict_);
+            modulo_ = std::move(other.modulo_);
+        }
+        return static_cast<GaloisFieldDict &>(*this);
+    }
+
+    GaloisFieldDict &operator+=(const GaloisFieldDict &other)
+    {
+        SYMENGINE_ASSERT(modulo_ == other.modulo_);
+        for (auto &iter : other.dict_) {
+            auto t = dict_.lower_bound(iter.first);
+            if (t != dict_.end() and t->first == iter.first) {
+                t->second += iter.second;
+                mp_fdiv_r(t->second, t->second, modulo_);
+                if (t->second == integer_class(0)) {
+                    dict_.erase(t);
+                }
+            } else {
+                integer_class temp;
+                mp_fdiv_r(temp, iter.second, modulo_);
+                dict_.insert(t, {iter.first, temp});
+            }
+        }
+        return static_cast<GaloisFieldDict &>(*this);
+    }
+
+    GaloisFieldDict operator-() const
+    {
+        GaloisFieldDict o(*this);
+        for (auto &a : o.dict_) {
+            a.second *= -1;
+            a.second += modulo_;
+        }
+        return o;
+    }
+
+    GaloisFieldDict &operator-=(const GaloisFieldDict &other)
+    {
+        SYMENGINE_ASSERT(modulo_ == other.modulo_);
+        for (auto &iter : other.get_dict()) {
+            auto t = dict_.lower_bound(iter.first);
+            if (t != dict_.end() and t->first == iter.first) {
+                mp_fdiv_r(t->second, t->second - iter.second, modulo_);
+                if (t->second == 0) {
+                    dict_.erase(t);
+                }
+            } else {
+                integer_class temp;
+                temp = modulo_ - iter.second;
+                mp_fdiv_r(temp, temp, modulo_);
+                dict_.insert(t, {iter.first, temp});
+            }
+        }
+        return static_cast<GaloisFieldDict &>(*this);
+    }
+
+    GaloisFieldDict &operator*=(const GaloisFieldDict &other)
+    {
+        SYMENGINE_ASSERT(modulo_ == other.modulo_);
+        if (dict_.empty())
+            return static_cast<GaloisFieldDict &>(*this);
+
+        auto o_dict = other.get_dict();
+        if (o_dict.empty()) {
+            *this = other;
+            return static_cast<GaloisFieldDict &>(*this);
+        }
+
+        // ! other is a just constant term
+        auto a = o_dict.find(0);
+        if (o_dict.size() == 1
+            and a != o_dict.end()) {
+            for (auto &arg : dict_) {
+                arg.second *= a->second;
+                mp_fdiv_r(arg.second, arg.second, modulo_);
+            }
+            return static_cast<GaloisFieldDict &>(*this);
+        }
+
+        map_uint_mpz dict_out;
+        for (auto &iter : o_dict) {
+            for (auto &it : dict_) {
+                mp_addmul(dict_out[iter.first + it.first], iter.second, it.second);
```

Comment:
Add `auto t = dict_out[iter.first + it.first];` above this line, and use it below.